crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+    },
+    validator_txn::{Topic, ValidatorTransaction},
+};
+use aptos_validator_transaction_pool::{TxnGuard, VTxnPoolState};
+use futures_channel::oneshot;
+use futures_util::{
+    future::{join_all, AbortHandle},
+    FutureExt, StreamExt,
+};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::Duration,
+};
+
+/// `JWKManager` executes per-issuer JWK consensus sessions
+/// and updates validator txn pool with quorum-certified JWK updates.
+pub struct JWKManager {
+    /// Some useful metadata.
+    my_addr: AccountAddress,
+    epoch_state: Arc<EpochState>,
+
+    /// Used to sign JWK observations before sharing them with peers.
+    consensus_key: Arc<PrivateKey>,
+
+    /// The sub-process that collects JWK updates from peers and aggregate them into a quorum-certified JWK update.
+    certified_update_producer: Arc<dyn CertifiedUpdateProducer>,
+
+    /// When a quorum-certified JWK update is available, use this to put it into the validator transaction pool.
+    vtxn_pool: VTxnPoolState,
+
+    /// The JWK consensus states of all the issuers.
+    states_by_issuer: HashMap<Issuer, PerProviderState>,
+
+    /// Whether a CLOSE command has been received.
+    stopped: bool,
+
+    qc_update_tx: Option<aptos_channel::Sender<(), QuorumCertifiedUpdate>>,
+    jwk_observers: Vec<JWKObserver>,
+}
+
+impl JWKManager {
+    pub fn new(
+        consensus_key: Arc<PrivateKey>,
+        my_addr: AccountAddress,
+        epoch_state: Arc<EpochState>,
+        certified_update_producer: Arc<dyn CertifiedUpdateProducer>,
+        vtxn_pool: VTxnPoolState,
+    ) -> Self {
+        Self {
+            consensus_key,
+            my_addr,
+            epoch_state,
+            certified_update_producer,
+            vtxn_pool,
+            states_by_issuer: HashMap::default(),
+            stopped: false,
+            qc_update_tx: None,
+            jwk_observers: vec![],
+        }
+    }
+
+    pub async fn run(
+        mut self,
+        oidc_providers: Option<SupportedOIDCProviders>,
+        observed_jwks: Option<ObservedJWKs>,
+        mut jwk_updated_rx: aptos_channel::Receiver<(), ObservedJWKsUpdated>,
+        mut rpc_req_rx: aptos_channel::Receiver<(), (AccountAddress, IncomingRpcRequest)>,
+        close_rx: oneshot::Receiver<oneshot::Sender<()>>,
+    ) {
+        self.reset_with_on_chain_state(observed_jwks.unwrap_or_default().into_providers_jwks())
+            .unwrap();
+        let (qc_update_tx, mut qc_update_rx) = aptos_channel::new(QueueStyle::FIFO, 100, None);
+        self.qc_update_tx = Some(qc_update_tx);
+
+        let (local_observation_tx, mut local_observation_rx) =
+            aptos_channel::new(QueueStyle::KLAST, 100, None);
+
+        self.jwk_observers = oidc_providers
+            .unwrap_or_default()
+            .into_provider_vec()
+            .into_iter()
+            .map(|provider| {
+                JWKObserver::spawn(
```

Comment:
Don't call it spawn.. confusing with tokio::spawn

Reply:
i believe it's actually the same idea? start something in the background and return some proxy object?